Source/missiles.cpp

```diff
@@ -2075,32 +2069,40 @@ void AddWeapexp(Missile &missile, Point dst, Direction /*midir*/)
 
 void AddTown(Missile &missile, Point dst, Direction /*midir*/)
 {
-	Point target = dst;
-	if (currlevel != 0) {
-		missile._miDelFlag = true;
-		for (int i = 0; i < 6; i++) {
-			int k = CrawlNum[i];
-			int ck = k + 2;
-			for (auto j = static_cast<uint8_t>(CrawlTable[k]); j > 0; j--, ck += 2) {
-				target = dst + Displacement { CrawlTable[ck - 1], CrawlTable[ck] };
-				if (!InDungeonBounds(target))
-					continue;
-
-				int dp = dPiece[target.x][target.y];
-				if (!TileContainsMissile(target) && !nSolidTable[dp] && !nMissileTable[dp] && dObject[target.x][target.y] == 0 && dPlayer[target.x][target.y] == 0) {
-					if (!CheckIfTrig(target)) {
-						missile.position.tile = target;
-						missile.position.start = target;
-						missile._miDelFlag = false;
-						i = 6;
-						break;
-					}
-				}
-			}
-		}
+	if (currlevel == 0) {
+		missile.position.tile = dst;
+		missile.position.start = dst;
 	} else {
-		missile.position.tile = target;
-		missile.position.start = target;
+		std::optional<Point> targetPosition = FindClosestValidPosition(
+		    [](Point target) {
+			    if (!InDungeonBounds(target)) {
+				    return false;
+			    }
+			    if (dObject[target.x][target.y] != 0) {
+				    return false;
+			    }
+			    if (dPlayer[target.x][target.y] != 0) {
+				    return false;
+			    }
+			    if (TileContainsMissile(target)) {
+				    return false;
+			    }
+
+			    int dp = dPiece[target.x][target.y];
+			    if (nSolidTable[dp] || nMissileTable[dp]) {
+				    return false;
+			    }
+			    return !CheckIfTrig(target);
+		    },
+		    dst, 0, 5);
+
+		if (targetPosition) {
+			missile.position.tile = *targetPosition;
+			missile.position.start = *targetPosition;
+			missile._miDelFlag = false;
+		} else {
+			missile._miDelFlag = true;

```

Comment:
nit:
Should we update/set `missile.position.tile` here?
In the old logic `dst` was used for `NetSendCmdLocParam3` and now we use `missile.position.tile`. I'm not sure if both are the same.
nit cause I'm not sure if we ever hit this case or if it's relevant, cause `_miDelFlag` is set to `true`.

Reply:
CMD_ACTIVATEPORTAL is only sent if _miDelFlag is false, in which case missile.position.tile has been set to target. PutMissile doesn't modify position (but may set delFlag to true) so I felt it was safe (and more descriptive) to show that the portal is being spawned where the missile is currently located.